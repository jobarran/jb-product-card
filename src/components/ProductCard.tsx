import React from "react";

import { useProduct } from '../hooks/useProduct'
import { createContext, CSSProperties } from 'react';
import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';

import styles from '../styles/styles.module.css'



export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

// Defino las props que voy a recibir
export interface Props {
    product: Product;
    // children?: ReactElement | ReactElement[];
    children: ( args: ProductCardHandlers ) => JSX.Element;
    className?: string;
    style?: CSSProperties;
    onChange?: ( args: onChangeArgs ) => void;
    value?: number;
    initialValues?: InitialValues
}

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }:Props) => {

    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues })

    return (

    <Provider value={{
        counter,
        increaseBy,
        maxCount,
        product
    }}>
        <div
            className={ `${styles.productCard} ${className}` }
            style={ style }
        >

            { 
                children({
                    count: counter,
                    isMaxCountReached: isMaxCountReached,
                    maxCount: initialValues?.maxCount,
                    product: product,
                    increaseBy: increaseBy,
                    reset: reset
                })
            }

            {/* <ProductImage img={ product.img } />
            <ProductTitle title={product.title } />
            <ProductButtons 
                counter={ counter }
                increaseBy={ increaseBy }
            /> */}

        </div>
    </Provider>

    )
}


