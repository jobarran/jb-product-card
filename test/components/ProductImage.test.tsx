import React from "react";
import renderer from "react-test-renderer";
import { ProductCard, ProductImage } from '../../src/components';
import { product2 } from '../../src/components/data/products';

describe('ProductImage', () => { 
    
    test('Debe mostrar el componente correctamente con la imagen personalizado', () => {

        const wrapper = renderer.create(
            <ProductImage 
                img="https://hola.jpg"
                className="custom-class"
            />
        )

        expect( wrapper.toJSON() ).toMatchSnapshot()


    });

    test('Debe mostrar el componente con la imagen del producto', () => {

        const wrapper = renderer.create(
            <ProductCard product={ product2 }>

            {
                () => (
                    <ProductImage />
                )
            }

            </ProductCard>
        )

        expect( wrapper.toJSON() ).toMatchSnapshot()

    })

 })