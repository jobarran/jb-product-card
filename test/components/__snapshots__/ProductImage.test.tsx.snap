// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe mostrar el componente con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product image"
    className="productCard undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`ProductImage Debe mostrar el componente correctamente con la imagen personalizado 1`] = `
<img
  alt="Product image"
  className="productCard custom-class"
  src="https://hola.jpg"
/>
`;
